cem/support-typedef-jsdoc.js

```diff
@@ -1,186 +1,119 @@
-        const fieldDoc = mb.jsDoc?.[0].tags?.[0].typeExpression;
-        // We need to have an "OR" (||) to find types who are object arrays e.g: Plan[] for example
-        const customType = fieldDoc?.type?.typeName?.getText() || fieldDoc?.type?.elementType?.typeName?.getText();
-
-        if (!isFieldPrivate && isFieldTypeUnion) {
-          fieldDoc.type.types.forEach((type) => {
-            // We need to have an "OR" (||) to find types who are object arrays e.g: Plan[] for example
-            const customType = type?.typeName?.getText() || type?.elementType?.typeName?.getText();
-            if (customType != null) {
-              types.push(customType);
-            }
-          });
-        }
-        else if (!isFieldPrivate && customType) {
-          types.push(customType);
-        }
-      });
+      moduleTypeCache.clear();
+      const rootDir = process.cwd();
+
+      // Now that we're in ClassDeclaration we find the constructor
+      const constructorNode = getConstructorNode(node, ts);
+
+      // Check if we have some component without constructor
+      if (constructorNode == null) {
+        return;
+      }
 
-      // Then, we loop through each element of the jsdoc in order to find the type imports
-      node?.jsDoc?.forEach((jsDoc) => {
+      console.log('⌛ Generating docs for', node.name.escapedText);
 
-        // For each tags we're gonna extract the file path of the "types.d.ts" file
-        // Then, we're gonna check that we have a matching type from the types we found previously
-        // in the constructor and add it to our imports set
-        jsDoc?.tags
-          ?.filter((tag) => tag.tagName.getText() === 'typedef' && (tag.typeExpression.type.kind === ts.SyntaxKind.ImportType))
-          ?.forEach((tag) => {
+      const types = getTypesFromConstructor(constructorNode, ts);
+      // Check the jsDoc of the class and find the imports
 
-            const moduleDir = path.parse(moduleDoc.path).dir;
-            // Remove leading and ending quotes
-            const typeRelativePath = tag.typeExpression.type.argument?.literal.getText().slice(1, -1);
-            const { dir: typeDir, name: typeName } = path.parse(typeRelativePath);
+      // This find the comment where the imports are located
+      const typeDefNode = node?.jsDoc?.filter((node) => node.tags.find((tag) => tag.kind === ts.SyntaxKind.JSDocTypedefTag))?.[0];
 
-            const typeToTs = path.format({ name: typeName, ext: '.d.ts' });
-            const typePath = path.resolve(rootDir, moduleDir, typeDir, typeToTs);
+      typeDefNode?.tags
+        ?.forEach((tag) => {
 
-            const typeDefDisplay = tag.name.getText();
-            const type = types.find((type) => type === typeDefDisplay);
+          // Detect if an import is not correct and warn the plugin user.
+          if (tag.kind !== ts.SyntaxKind.JSDocTypedefTag) {
+            console.log('Something in your @typedef imports is wrong.');
```

Comment:
You can know what component is faulty because the plugin tells you before that phase which component documentation is being generated.
Nonetheless we should be able to print the faulty line as we do in the script that checks type imports. :wink: